### PR TITLE
Send an immediate ACK

### PR DESCRIPTION
**Pull request by ianswett**
Attempt to address #1968

### CODE REVIEW

File: draft-ietf-quic-recovery.md [hunk 1]
```diff
@@ -483,11 +483,15 @@ the peer spurious timeouts.  The maximum ack delay is communicated in the
 An acknowledgement SHOULD be sent immediately upon receipt of a second
 packet but the delay SHOULD NOT exceed the maximum ack delay. QUIC recovery
 algorithms do not assume the peer generates an acknowledgement immediately when
```
Comment (mikkelfj) [inline draft-ietf-quic-recovery.md]: does immediately mean ASAP or any convenient time before ACK delay ?

File: draft-ietf-quic-recovery.md [hunk 2]
```diff
@@ -483,11 +483,15 @@ the peer spurious timeouts.  The maximum ack delay is communicated in the
...
-receiving a second full-packet.
+receiving a second packet.
 
 Out-of-order packets SHOULD be acknowledged more quickly, in order to accelerate
 loss recovery.  The receiver SHOULD send an immediate ACK when it receives a new
-packet which is not one greater than the largest received packet number.
+packet which is not one greater than the largest received packet number.  The
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: Suggestion: "The receiver SHOULD send an immediate ACK when it receives a new packet which is not the next expected one.  That is, its packet number is not one greater than the largest received packet number."

File: draft-ietf-quic-recovery.md [hunk 3]
```diff
@@ -483,11 +483,15 @@ the peer spurious timeouts.  The maximum ack delay is communicated in the
...
+receiver MAY send an immediate ACK when a missing packet begins exceeding the
+peer's reordering threshold in packets to accelerate fast retransmit.  Time
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: This is a problem, since the receiver does not know the peer's reordering threshold. Also, this sentence does not make sense "a missing packet begins exceeding ... threshold in packets"

File: draft-ietf-quic-recovery.md [hunk 4]
```diff
@@ -483,11 +483,15 @@ the peer spurious timeouts.  The maximum ack delay is communicated in the
...
+threshold loss detection does not require any special ACK generation to
+accelerate fast retransmit, because it uses a timer.
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: I think this text should be purely prescriptive, and only from the receiver's point of view, since the sender's is basically not known at the receiver. This also does not say when a receiver should stop immediate acking, which is important IMO.

Suggested replacement for text from "The receiver MAY" : "A receiver SHOULD immediately ack at least three subsequent packets after first receiving a packet out of order, after which it SHOULD resume delaying acknowledgements."

**Comment by martinduke** [comment]
I think 'kPacketThreshold' is the ideal value of 'few' in my comment above. Maybe this should become an optional transport parameter?

File: draft-ietf-quic-recovery.md [hunk 5]
```diff
@@ -224,11 +224,16 @@ the peer spurious timeouts.  The maximum ack delay is communicated in the
...
-receiving a second full-packet.
+receiving a second packet.
 
 Out-of-order packets SHOULD be acknowledged more quickly, in order to accelerate
 loss recovery.  The receiver SHOULD send an immediate ACK when it receives a new
-packet which is not one greater than the largest received packet number.
+packet which is not the next expected one. That is, its packet number is not one
+greater than the largest received packet number.  A receiver MAY immediately
+ack subsequent packets after first receiving a packet out of order. A receiver
+SHOULD NOT send an immediate ACK any time there is a gap in the ACK frame,
```
[the remark this answers is not in the document]
Comment (ianswett) [reply draft-ietf-quic-recovery.md]: Clarified.

**Comment by martinduke** [comment]
@ianswett good point; I return to the position of my comment before last.

File: draft-ietf-quic-recovery.md [hunk 6]
```diff
@@ -222,13 +222,17 @@ the peer spurious timeouts.  The maximum ack delay is communicated in the
...
 
 Out-of-order packets SHOULD be acknowledged more quickly, in order to accelerate
 loss recovery.  The receiver SHOULD send an immediate ACK when it receives a new
-packet which is not one greater than the largest received packet number.
+packet which is not the next expected one. That is, its packet number is not one
+greater than the largest received packet number.  A receiver MAY immediately
+acknowledge subsequent ack-eliciting packets after first receiving a packet out
+of order. A receiver SHOULD NOT send an immediate ACK any time there is a gap
+in the ACK frame being sent, because that will cause it to send an ACK in
+response to every packet for at least an RTT after a single packet loss.
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: I don't understand what the text here is trying to say.  I'll suggest rewording for text "A receiver MAY ... single packet loss", see if this matches what you're getting at.

"A receiver MAY send acknowledgements immediately for the next few packets that are received, but the receiver SHOULD revert back to delaying acknowledgements after sending a few immediate acknowledgements."

File: draft-ietf-quic-recovery.md [hunk 7]
```diff
@@ -222,13 +222,16 @@ the peer spurious timeouts.  The maximum ack delay is communicated in the
...
-packet but the delay SHOULD NOT exceed the maximum ack delay. QUIC recovery
-algorithms do not assume the peer generates an acknowledgement immediately when
-receiving a second full-packet.
+ack-eliciting packet. QUIC recovery algorithms do not assume the peer sends an
+an ACK immediately when receiving a second ack-eliciting packet.
 
 Out-of-order packets SHOULD be acknowledged more quickly, in order to accelerate
 loss recovery.  The receiver SHOULD send an immediate ACK when it receives a new
-packet which is not one greater than the largest received packet number.
+packet which is not the next expected one. That is, its packet number is not one
```
Comment (martinduke) [inline draft-ietf-quic-recovery.md]: Again, there are SHOULDs on top of SHOULDs here. How about this for a paragraph:

"Quicker acknowledgement of out-of-order packets may accelerate loss detection. Upon receipt of a packet number not one more than the previous maximum received, receivers SHOULD separately acknowledge it and the next few packets, but not more than five."

This is shorter, has only one SHOULD, and places a recommended limit that is less subjective than "a few".

Comment (ianswett) [reply draft-ietf-quic-recovery.md]: Thanks for the suggestions, I reworked the text to only have one SHOULD and mention the 1/8 RTT as the max amount of time to send immediate acks.
